fv3core/stencils/updatedzc.py

```diff
@@ -1,214 +1,72 @@
-#            do i=is1, ie1
-#               yfx(i,j) = (dp0(k)*vt(i,j,k-1)+dp0(k-1)*vt(i,j,k))*int_ratio
-#            enddo
-#         enddo
-#      endif
-
-
 @gtscript.function
 def p_weighted_average_domain(vel, dp0):
-    # ratio = dp0 / ( dp0[0, 0, -1] + dp0 )
-    # return ratio * vel[0, 0, -1] + (1. - ratio) * vel
     int_ratio = 1.0 / (dp0[0, 0, -1] + dp0)
     return (dp0 * vel[0, 0, -1] + dp0[0, 0, -1] * vel) * int_ratio
 
 
-#      do j=js-ng, je+ng
-#         do i=is-ng, ie+ng
-#            gz2(i,j) = gz(i,j,k)
-#         enddo
-#      enddo
-
-#      if (grid_type < 3) call fill_4corners(gz2, 1, bd, npx, npy, sw_corner, se_corner, ne_corner, nw_corner) # noqa: E501
-
-#      do j=js1, je1
-#         do i=is1, ie2
-#            if( xfx(i,j) > 0.0 ) then
-#                fx(i,j) = gz2(i-1,j)
-#            else
-#                fx(i,j) = gz2(i  ,j)
-#            endif
-#            fx(i,j) = xfx(i,j)*fx(i,j)
-#         enddo
-#      enddo
-
-#      if (grid_type < 3) call fill_4corners(gz2, 2, bd, npx, npy, sw_corner, se_corner, ne_corner, nw_corner) # noqa: E501
-
-#      do j=js1,je2
-#         do i=is1,ie1
-#            if( yfx(i,j) > 0.0 ) then
-#                fy(i,j) = gz2(i,j-1)
-#            else
-#                fy(i,j) = gz2(i,j)
-#            endif
-#            fy(i,j) = yfx(i,j)*fy(i,j)
-#         enddo
-#      enddo
-
-
 @gtscript.function
 def xy_flux(gz_x, gz_y, xfx, yfx):
     fx = xfx * (gz_x[-1, 0, 0] if xfx > 0.0 else gz_x)
     fy = yfx * (gz_y[0, -1, 0] if yfx > 0.0 else gz_y)
     return fx, fy
 
 
-#      do j=js1, je1
-#         do i=is1,ie1
-#            gz(i,j,k) = (gz2(i,j)*area(i,j) +  fx(i,j)- fx(i+1,j)+ fy(i,j)- fy(i,j+1)) & # noqa: E501
-#                      / (         area(i,j) + xfx(i,j)-xfx(i+1,j)+yfx(i,j)-yfx(i,j+1)) # noqa: E501
-#         enddo
-#      enddo
-
-#   end do
-
-# ! Enforce monotonicity of height to prevent blowup
-#   do i=is1, ie1
-#      do j=js1, je1
-#         ws(i,j) = ( zs(i,j) - gz(i,j,km+1) ) * rdt
-#      enddo
-#   end do
-
-#   do k = km, 1, -1
-#      do j=js1, je1
-#         do i=is1, ie1
-#            gz(i,j,k) = max( gz(i,j,k), gz(i,j,k+1) + dz_min )
-#         enddo
-#      enddo
-#   enddo
-
-
 @gtstencil()
-def update_dz_c(
-    dp_ref: sd,
-    zs: sd,
-    area: sd,
-    ut: sd,
-    vt: sd,
-    gz: sd,
-    gz_x: sd,
-    gz_y: sd,
-    ws3: sd,
-    *,
-    dt: float,
+def update_dz_c_stencil(
+    area: FloatField,
+    dp_ref: FloatField,
+    z_surface: FloatField,
+    ut: FloatField,
+    vt: FloatField,
+    gz: FloatField,
+    ws: FloatField,
+    dt2: float,
 ):
+    """Update the model heights from the C-grid wind flux
+
+    After the model runs c_sw and advances the c-grid variables half a timestep,
+    the grid deforms with the flow in the Lagrangian coordinate system. This
+    module updates the model heights based on the flux, and the rate of change of
+    the lowest model height compared to the fixed surface height.
```

Comment:
Do you know from what rule/assumptions it's able to update the heights? It's not hydrostatic, so I am having a hard time imagining what constrains the height change.

If you know, it would be great to add here. If not, I think Lucas would be happy to receive a question about this.

Reply:
Yes, that's definitely true/accurate, but it's not the full story. I couldn't from that information write a formula (or have any idea how to start) that would take the winds and determine how much vertical expansion/contraction there is.

If this were an incompressible model, I could do it by using conservation of volume - based on how much horizontal divergence (convergence) there is, I could say how much contraction (expansion) of the column there is in the vertical. With a hydrostatic assumption, I could make a similar statement.

For this model, I have no idea what the assumption is that gives closure. As far as I know, all the horizontal divergence/convergence could go to increasing delp/density while keeping dz entirely constant.

Do you have any idea what the closure assumption is here?